odl/contrib/tensorflow.py

```diff
@@ -0,0 +1,512 @@
+
+    def py_func(func, inp, Tout, stateful=True, name=None, grad=None):
+        """Define custom py_func which takes also a grad op as argument.
+
+        We need to overwrite this function since the default tensorflow py_func
+        does not support custom gradients.
+
+        Parameters
+        ----------
+        func : callable
+            Python function that takes and returns numpy arrays.
+        inp : sequence of `tensorflow.Tensor`
+            Input tensors for the function
+        Tout : sequence of `tensorflow.dtype`
+            Datatype of the output(s).
+        stateful : bool
+            If the function has internal state, i.e. if calling the function
+            with a given input always gives the same output.
+        name : string
+            Name of the python function
+        grad : callbable
+            Gradient of the function.
+        """
+        if grad is None:
+            return tf.py_func(func, inp, Tout, stateful=stateful, name=name)
+        else:
+            if stateful:
+                override_name = 'PyFunc'
+            else:
+                override_name = 'PyFuncStateless'
+
+            # Need to generate a unique name to avoid duplicates:
+            rnd_name = override_name + 'Grad' + str(uuid.uuid4())
+
+            tf.RegisterGradient(rnd_name)(grad)
+            g = tf.get_default_graph()
+
+            with g.gradient_override_map({override_name: rnd_name}):
+                return tf.py_func(func, inp, Tout, stateful=stateful,
+                                  name=name)
+
+    def tensorflow_layer_grad_impl(x, dx, name):
+        """Implementation of the tensorflow gradient.
+
+        Gradient in tensorflow is equivalent to the adjoint of the derivative
+        in ODL.
+
+        Returns a `tensorflow.Tensor` that represents a lazy application of ::
+
+            odl_op.derivative(x).adjoint(dx)
+
+        Parameters
+        ----------
+        x : `numpy.ndarray`
+            Point(s) in which the derivative should be taken.
+            If ``odl_op`` is an `Operator` the axes are:
+                0 : batch id. This is a constant if ``fixed_size`` is
+                    true, otherwise it is dynamic.
+                1, ..., N-2 : spatial dimensions of data.
+                n-1 : (currently) unused data channel.
+            If ``odl_op`` is a `Functional` the axes are:
+                0 : batch id.
+        dx : `tensorflow.Tensor`
+            Point(s) in which the adjoint of the derivative of the
+            operator should be evaluated.
+            The axes are:
+                0 : batch id. Should be pairwise matched with ``x``.
+                1, ..., M-2 : spatial dimensions of data.
+                n-1 : (currently) unused data channel.
+        name : string
+            Name of the tensor.
+
+        Returns
+        -------
+        result : `tensorflow.Tensor`
+            Lazy result of the computation.
```

Comment:
I guess it is similar to Theano's `function` where you do (roughly):
```python
x = TensorVariable(...)
y = <some expression in x>
f = function(x, y)  # nothing computed
f(np.zeros(...))  # now evaluated
```
The `function(x, y)` thing builds the computation graph, and `x` and `y` are symbolic variables.